### PR TITLE
Cardinality API: add the ability to filter by "active" series

### LINKED ISSUE #5135: Cardinality API: Support filtering by active series
#### Is your feature request related to a problem? Please describe.

Users of Grafana Cloud are billed based on the concept of "active series". Currently the cardinality API returns the number of series in the tsdb head block, which can be wildly different when series churn is present. 

#### Describe the solution you'd like

It would be more useful if the cardinality API could return only active series. Both numbers are useful, so I would suggest making the "active series" number an optional filter.

#### Describe alternatives you've considered

Customers can get the same information using the promql query `count(present_over_time(<selector>[20m]))`, but this is much slower when compared to the cardinality API, and requires loading chunks from the tsdb.

#### TODO

- [ ] https://github.com/grafana/mimir/pull/5134
- [ ] #5136

### PR DESCRIPTION
**Pull request by Logiraptor**
#### What this PR does

This PR builds on #5134 to support a concept of `count_method` for the label values cardinality API. This is a new query parameter on the `/api/v1/cardinality/label_values` endpoint with two possible values:

1. `inmemory`: the default, current behavior. This counts series that exist in the tsdb head block
2. `active`: new behavior. This counts series from the tsdb head block that are also "active" according to the active series tracker.

Here's an example from a small local Mimir installation shortly after rolling out an update: 

Screenshot 2023-06-01 at 5 38 01 PM

```sh
curl -v -H "X-Scope-OrgID: <tenant>" "http://localhost:8080/prometheus/api/v1/cardinality/label_values" -d 'label_names[]=__name__' -d 'count_method=inmemory'
```

```json
{
  "series_count_total": 53041,
  "labels": [
    {
      "label_name": "__name__",
      "label_values_count": 705,
      "series_count": 53041,
      "cardinality": [
        {
          "label_value": "thanos_objstore_bucket_operation_duration_seconds_bucket",
          "series_count": 7245
        },
        {
          "label_value": "cortex_request_duration_seconds_bucket",
          "series_count": 3915
        }
       ]
     }
   ]
}
```

vs the new behavior:

```sh
curl -v -H "X-Scope-OrgID: <tenant>" "http://localhost:8080/prometheus/api/v1/cardinality/label_values" -d 'label_names[]=__name__' -d 'count_method=active'
```

```json
{
  "series_count_total": 17747,
  "labels": [
    {
      "label_name": "__name__",
      "label_values_count": 689,
      "series_count": 17747,
      "cardinality": [
        {
          "label_value": "thanos_objstore_bucket_operation_duration_seconds_bucket",
          "series_count": 2415
        },
        {
          "label_value": "cortex_request_duration_seconds_bucket",
          "series_count": 1320
        }
      ]
    }
  ]
}
```

Normal series churn during the rollout means the in memory series count is around 2-3x the active series count. With this change, users can now see both values.

Submitting this PR now to get early feedback before finishing the tests and docs. Please let me know what you think!

~~I went back and forth on a couple designs for the API, but I'm not totally sold on the naming of `scope` as a query parameter, so I'm happy to hear alternative designs.~~ We settled on `count_method` as the parameter name.

#### Which issue(s) this PR fixes or relates to

Fixes #5135 

#### Checklist

- [X] Tests updated
- [X] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by Logiraptor** [comment]
Notable limitations of this implementation:

In this response:
```
{
  "series_count_total": 53041,
  "labels": [
    {
      "label_name": "__name__",
      "label_values_count": 689,
      "series_count": 17747,
      "cardinality": [
        {
          "label_value": "thanos_objstore_bucket_operation_duration_seconds_bucket",
          "series_count": 2415
        },
        {
          "label_value": "cortex_request_duration_seconds_bucket",
          "series_count": 1320
        }
      ]
    }
  ]
}
```

* The top level `series_count_total` still reports total series in the tsdb head. It uses a different mechanism which isn't so easy to filter
* The `label_values_count` parameter is also based on the in memory series. This means you could have results where `series_count` is `0`, because the metric existed at some earlier point in the tsdb head, but is no longer active.

For me, these seem like acceptable tradeoffs in order to give users the _main value_, which is accurate values in all the `series_count` fields. With this, users can accurately measure the impact of changes in ~realtime while trying to optimize their spend on Grafana Cloud.

I've chosen to leave the `inmemory` result as a default, because it is more useful for OSS users operating Mimir to understand which label values are contributing the most to ingester memory.

**Comment by RutgerKe** [comment]
> The top level series_count_total still reports total series in the tsdb head. It uses a different mechanism which isn't so easy to filter

I'm very concerned that the fact that these `series_count`  values don't sum to the `series_count_total` will cause a lot of confusion. I think the main use case for the `series_count_total` value is to compute the % a `series_count` attributes to the total, and this can be under-reporting up to 2-3x  (so lets say: a metric looks like 10%, but actually contributes 25%). I would also not trust the results if a simple 'sum all parts and compare to the total' doesn't hold.

You might be ahead of me; I don't know why its hard to filter that number to active, so sorry if you've already considered this, but my thought:

If I got it correctly, this is the total regardless of the extra parameters to filter down results, correct? In that case: Isn't this the same as the 'active series' metric we expose per tenant (which we use for billing purposes?), can't we surface that same value here?

(If its the total of the response than of we could simply sum the `series_count`s we see in the response before returning it?)

**Comment by dimitarvdimitrov** [comment]
I hadn't read the comments and PR description. I think it's important to give a consistent response to the user. Now the number of total series seems to be computed via a different RPC. So there's still a race condition between the RPCs for LabelValues and for UserStats, but it won't be 2x,3x as Rutger points out.

File: pkg/querier/distributor_queryable.go
Comment (colega) [inline pkg/querier/distributor_queryable.go:33]: Nit: while having the type in the name is arguable, to avoid possible conflicts, I think it's better to start naming with the meaningful part, and then the type-deduplication part, because it's faster for the brain to understand `querier.InMemorySeriesScope` than `querier.SeriesScopeInMemory`.
And I'd drop the `Series` part too, making it just `querier.InMemoryScope`.

---

_Now, if this was my codebase_, I would actually take advantage of golang's package.type naming, and would have a:

```go
package seriesscope

type Scope string

const (
	InMemory Scope = "inmemory"
	Active   Scope = "active"
)
```

So the usages would be: `seriesscope.InMemory`.

However, this seems to go against our usual pattern of _"put as many things as possible in a single package"._

**Comment by colega** [comment]
> The top level series_count_total still reports total series in the tsdb head. It uses a different mechanism which isn't so easy to filter

I don't see it so hard to filter, it's actually easier than what you did.

As @RutgerKe says:

> If I got it correctly, this is the total regardless of the extra parameters to filter down results, correct? In that case: Isn't this the same as the 'active series' metric we expose per tenant, can't we surface that same value here?

So we just need to pass down the `ActiveSeriesOnly` flag in the `UserStats` request we're sending, and then here:
https://github.com/grafana/mimir/blob/fef8cd465e6b43fdd63b7766b463e16439a5dab5/pkg/ingester/ingester.go#L1377-L1377

Just use the result of `db.activeSeries.Active(time.Now())`.
(As an extra optimization, I'd also pass to that function a flag (or create a different function) telling it that we only want the active series, we don't want the active series for custom matchers).

**Comment by Logiraptor** [comment]
@colega @dimitarvdimitrov @RutgerKe I see you're right we can easily filter that value as well. Here's my hesitation with changing `series_count_total`:

The `series_count_total` value is meant to represent "total number of series across opened TSDBs in all ingesters" according to the docs.

We already have a `selector` parameter, which can be used to limit the series analyzed. When `selector` is set, the `series_count_total` does not take this into consideration, and the sum of `series_count` does not add up to the `series_count_total`.  It is _always_ the total series in memory. I see this new parameter as a similar sort of filter, and I would expect it to behave the same way. 

What do you think?

**Comment by colega** [comment]
It doesn't really convince me.
I agree that we're in a suboptimal state, but let's not do it worse: if the request is talking about active series, all numbers should refer to active series.

**Comment by Logiraptor** [comment]
There's two other scenarios where the `series_count_total` won't match the sum of `series_count`. 

1. If you group by a label that exists only on a subset of series
2. If there are more label values than the `limit`

So I'm wondering what sounds most reasonable:

1. The `series_count_total` is the total active series across all ingesters when querying active series, ignoring selector, limit, etc.
2. The `series_count_total` is the sum of `series_count`, and exists only as a shortcut so you don't have to add numbers up

I'm leaning towards 1

**Comment by Logiraptor** [comment]
I've updated `series_count_total` to reflect the active series across the cluster as requested.

**Comment by RutgerKe** [comment]
> I've updated `series_count_total` to reflect the active series across the cluster as requested.

Thanks! This is indeed what I think is best, leaving the reasoning I was going to write in a comment before I saw you changed it:

> We already have a selector parameter, which can be used to limit the series analyzed. When selector is set, the series_count_total does not take this into consideration, and the sum of series_count does not add up to the series_count_total. It is always the total series in memory. I see this new parameter as a similar sort of filter, and I would expect it to behave the same way.

Without the 'selector' set (and within limits), the counts _should_ sum up to the total.

I'd say a lot of requests will have the selector set, or return more than the limit the API returns. In that case this 'total' value is needed to compare the counts you do get back, to understand their values relative to the total. The "total number of series across opened TSDBs in all ingesters" served that purpose before, but is not useful to compare to when working with 'active series'.

So thank you for changing it :+1:

File: pkg/ingester/activeseries/active_series.go
Comment (colega) [inline pkg/ingester/activeseries/active_series.go:112]:
```suggestion
	c.mu.RLock()
	defer c.mu.RUnlock()
```

Comment (colega) [reply pkg/ingester/activeseries/active_series.go:112]: Actually, can you double or triple check whether we need to hold this lock, and what's the purpose of it? As far as I can see, we don't hold it in `UpdateSeries`, so what's the point in holding it here?

I have a feeling that it's more related to a possible matchers change.

**Review by colega — COMMENTED** [review]
LGTM in general, but let's make sure what's going on with that mutex before merging.

File: pkg/ingester/activeseries/active_series.go
Comment (dimitarvdimitrov) [reply pkg/ingester/activeseries/active_series.go:112]: each stripe also has an individual lock. I also think the global lock on ActiveSeries is only for config. We should probably rename it in that case

File: pkg/ingester/client/ingester.proto
Comment (dimitarvdimitrov) [inline pkg/ingester/client/ingester.proto:57]: Will this change based on the survey you opened? I know the survey is for the HTTP API, but now we have the chance to make gRPC and HTTP align.

Comment (dimitarvdimitrov) [reply pkg/ingester/client/ingester.proto:57]: Same question applies for `UserStatsRequest`?

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM with one question about the proto layout

File: pkg/ingester/client/ingester.proto
Comment (Logiraptor) [reply pkg/ingester/client/ingester.proto:57]: Sure, I think it's a good idea to make them match, I'm planning to make a final decision tomorrow, then I'll take a last look at the grpc api to make sure it makes sense.

**Comment by 09jvilla** [comment]
I voted in the poll but I liked /api/v1/cardinality/label_values?count_method. I also suggested /api/v1/cardinality/label_values?count_strategy.

**Comment by 09jvilla** [comment]
I initially had some feedback on the choice of "active" series as a name (instead thinking maybe "billable" made more sense?) 

> 'm worried people might get confused by Prometheus which has an active series concept as well (and I don't think these two are the same but maybe I'm wrong?)
A series with an open head chunk is called an active series. This is different from a memory series, which also includes series without an open head chunk but still other chunks in memory (whether waiting for persistence, used in a query, or evictable)
Either way, I think the question is "if I'm an OSS Mimir operator, would I ever care about 'active series'?" If the answer is no, it would be ideal to name this in such a way that it made that very obvious (I'd argue its not as obvious right now because its using a term from OSS prometheus). Even better, perhaps this doesn't make sense to have this count method in the OSS at all.
I initially thought this 'active series' count wouldn't be relevant to a person operating GEM either, but on further thought, I guess it actually is since we've now moved GEM to be billed on active series, and we're going to want to use the same unit of active series as we do for Grafana Cloud.

https://raintank-corp.slack.com/archives/C9KL8THCY/p1686021045310749?thread_ts=1685743494.013079&cid=C9KL8THCY

**Comment by colega** [comment]
@09jvilla note that [we have references to active series in the configuration](https://grafana.com/docs/mimir/latest/references/configuration-parameters/) and also to [active series custom trackers](https://grafana.com/docs/mimir/latest/references/configuration-parameters/). I think that if we wanted to call these `billable`, we'd have to rename all of the config/features.

> could I change the window of what is considered "active" to more closely line up with "in memory" series ?
It can be configured using `-ingester.active-series-metrics-idle-timeout` flag. But it would never really "line up" with in-memory series. Actually, I'm not sure that this feature would work properly once the idle-timeout reaches the block size.

**Comment by 09jvilla** [comment]
> @09jvilla note that [we have references to active series in the configuration](https://grafana.com/docs/mimir/latest/references/configuration-parameters/) and also to [active series custom trackers](https://grafana.com/docs/mimir/latest/references/configuration-parameters/). I think that if we wanted to call these billable, we'd have to rename all of the config/features.

Good point - yet another reason to leave as active. :)

**Review by colega — APPROVED** [review]

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [inline docs/sources/mimir/references/http-api/index.md:572]:
```suggestion
Two subsequent calls may return completely different results if an ingester cut a block between calls.
```

Is this the only case where this is true? Or is "block cutting between the calls" one example scenario (of many possibilities) which could cause this?

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [inline docs/sources/mimir/references/http-api/index.md:575]:
```suggestion
The `active` method also counts series in currently opened TSDBs in Mimir's ingesters, but filters out series that have not received a sample within a configurable duration of time. 
```

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [inline docs/sources/mimir/references/http-api/index.md:576]:
```suggestion
Use the `ingester.active-series-metrics-idle-timeout` parameter to configure this duration.
```

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [inline docs/sources/mimir/references/http-api/index.md:577]:
```suggestion
This method of counting is most useful for understanding which label values are represented in the samples ingested by Mimir in the last `ingester.active-series-metrics-idle-timeout` minutes.
```

I'm not necessarily sure that its minutes but does this phrasing work?

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [inline docs/sources/mimir/references/http-api/index.md:579]:
```suggestion
Values will change only as a result of changes in the data ingested by Mimir. 
```

Its not necessarily the rate right? If the rate goes up because I go from 4 dpm to 16 dpm but I'm still sending the same label set, I would expect the answer to remain the same right?

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [reply docs/sources/mimir/references/http-api/index.md:575]: honestly I still don't understand how if `active` is just a subset of `inmemory` (you said it "filters stuff out" iiuc) how its not also impacted by this same block cutting problem? But I don't necessarily think I need to understand this =)

**Review by 09jvilla — CHANGES_REQUESTED** [review]
Left a few comments - mostly just all grammatical but a few were me trying to test my understanding.

**Review by 09jvilla — CHANGES_REQUESTED** [review]
Left a few comments - mostly just all grammatical but a few were me trying to test my understanding.

**Review by 09jvilla — CHANGES_REQUESTED** [review]
Left a few comments - mostly just all grammatical but a few were me trying to test my understanding.

File: docs/sources/mimir/references/http-api/index.md
Comment (Logiraptor) [reply docs/sources/mimir/references/http-api/index.md]: Yep, the unit is specified in the value. So you can say `5m` to mean 5 minutes or `30s` to mean 30 seconds.

File: docs/sources/mimir/references/http-api/index.md
Comment (osg-grafana) [inline docs/sources/mimir/references/http-api/index.md:569]:
```suggestion
Two methods of counting are available: `inmemory` and `active`. To choose one, use the `count_method` parameter.
```

File: docs/sources/mimir/references/http-api/index.md
Comment (osg-grafana) [inline docs/sources/mimir/references/http-api/index.md:572]:
```suggestion
Two subsequent calls might return completely different results if an ingester cut a block between calls.
```

File: docs/sources/mimir/references/http-api/index.md
Comment (osg-grafana) [inline docs/sources/mimir/references/http-api/index.md:576]:
```suggestion
To configure this duration, use the `ingester.active-series-metrics-idle-timeout` parameter.
```

File: docs/sources/mimir/references/http-api/index.md
Comment (osg-grafana) [inline docs/sources/mimir/references/http-api/index.md:577]:
```suggestion
This method of counting is most useful for understanding what label values are represented in the samples ingested by Mimir in the last `ingester.active-series-metrics-idle-timeout`.
```

File: docs/sources/mimir/references/http-api/index.md
Comment (osg-grafana) [inline docs/sources/mimir/references/http-api/index.md:578]:
```suggestion
Two subsequent calls will likely return similar results, because this window of time is not related to the block cutting on ingesters.
```

**Review by osg-grafana — APPROVED** [review]
Unblocking with feedback

File: docs/sources/mimir/references/http-api/index.md
Comment (pracucci) [inline docs/sources/mimir/references/http-api/index.md:576]:
```suggestion
Use the `-ingester.active-series-metrics-idle-timeout` parameter to configure this duration.
```

File: docs/sources/mimir/references/http-api/index.md
Comment (pracucci) [inline docs/sources/mimir/references/http-api/index.md:577]:
```suggestion
This method of counting is most useful for understanding which label values are represented in the samples ingested by Mimir in the last `-ingester.active-series-metrics-idle-timeout`.
```

File: pkg/ingester/activeseries/active_series.go
Comment (pracucci) [inline pkg/ingester/activeseries/active_series.go:111]: What if we move the purge logic out of `ActiveWithMatchers()` too, into a dedicated function (which will be called after `ActiveWithMatcher()`? From a design perspective I think it's weird that `ActiveWithMatchers()` does the purge but `Active()` doesn't.

Comment (Logiraptor) [reply pkg/ingester/activeseries/active_series.go]: Good idea, done in https://github.com/grafana/mimir/pull/5136/commits/8a71746a41eb6c428e1191a469b64e48aaefc367

File: docs/sources/mimir/references/http-api/index.md
Comment (09jvilla) [reply docs/sources/mimir/references/http-api/index.md]: Thank you for explaining!

**Review by 09jvilla — APPROVED** [review]
LGTM.

File: CHANGELOG.md
Comment (Logiraptor) [inline CHANGELOG.md:9]:
```suggestion
* [CHANGE] Querier: change the default value of the experimental `-querier.streaming-chunks-per-ingester-buffer-size` flag to 256. #5203
* [FEATURE] Cardinality API: Add a new `count_method` parameter which enables counting active series #5136
```

**Pull request merged**
